### PR TITLE
Make world config dialog and advanced settings dialog support nested modpacks

### PR DESCRIPTION
**Pull request by orwell96**
The engine is able to handle nested modpack structures, while the menu was not.
The entry format of mod entries from get_mods was changed, see the comment on top of modmgr.lua for details

### CODE REVIEW

File: builtin/mainmenu/modmgr.lua [hunk 1]
```diff
@@ -31,17 +50,16 @@ function get_mods(path,retval,modpack)
...
-				toadd.modpack = modpack
-			else
-				local modpackfile = io.open(prefix .. "modpack.txt")
-				if modpackfile then
-					modpackfile:close()
-					toadd.is_modpack = true
-					get_mods(prefix, retval, name)
-				end
+			toadd.modpack = modpack
+			toadd.mp_level=mp_level
```
Comment (SmallJoker) [inline builtin/mainmenu/modmgr.lua]: Spaces

File: builtin/mainmenu/modmgr.lua [hunk 2]
```diff
@@ -15,10 +15,29 @@
...
 
+
+-- On the format of the mods table returned by get_mods:
+-- This is an example entry for the 'lrfurn' mod found in the homedecor modpack
+-- {
+--    name = "LivingRoomFurniture",
+--       human-readable name of the mod/modpack, the name written in mod.conf if available
+--    full_name = "/homedecor/lrfurn",
+--       pseudo-path to the mod in the modpack directory structure. This field is the sort criteria in sort_mod_list()
+--    path = "/home/uset/.minetest/mods/homedecor/lrfurn",
```
Comment (SmallJoker) [inline builtin/mainmenu/modmgr.lua]: Typo in "user"

**Comment by orwell96** [comment]
Changes applied.
I didn't do a block comment because the license note above wasn't a block comment either.
closing with ]]-- is just because it looks nice

**Comment by orwell96** [comment]
!! wait !! I need to correct something!

**Comment by orwell96** [comment]
If there are modpacks inside the game, they don't get sorted into the tree but just printed one after another. I'm fixing this, meanwhile don't merge please!

**Comment by orwell96** [comment]
OK, done.

**Comment by orwell96** [comment]
OK, done.

File: mods/mods_here.txt [hunk 3]
```diff
@@ -1,4 +0,0 @@
-You can install Minetest mods by copying (and extracting) them into this folder.
-To enable them, go to the configure world window in the main menu or write
-  load_mod_<modname> = true
-in world.mt in the world directory.
```
Comment (SmallJoker) [inline mods/mods_here.txt]: If you remove this file, the whole mods directory will disappear aswell. Shouldn't we at least leave this file existent to keep the structure?

File: builtin/mainmenu/modmgr.lua [hunk 4]
```diff
@@ -15,10 +15,30 @@
...
 --51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 
+--[[
+ On the format of the mods table returned by get_mods:
+ This is an example entry for the 'lrfurn' mod found in the homedecor modpack
+ {
+    name = "LivingRoomFurniture",
+       human-readable name of the mod/modpack, the name written in mod.conf if available
+    full_name = "/homedecor/lrfurn",
+       pseudo-path to the mod in the modpack directory structure. This field is the sort criteria in sort_mod_list()
```
Comment (SmallJoker) [inline builtin/mainmenu/modmgr.lua]: Line too long (>80 chars)

File: builtin/mainmenu/modmgr.lua [hunk 5]
```diff
@@ -15,10 +15,30 @@
...
+    path = "/home/user/.minetest/mods/homedecor/lrfurn",
+       filesystem path to the mods/modpack's base directory
+    modpack = "/homedecor",
+       full_name of the parent modpack. If mod is not in any modpack, value is "".
+    mp_level = 1
+       Intendation level where this mod(pack) should be drawn in a tree. 0 is top layer (modpack "")
```
Comment (SmallJoker) [inline builtin/mainmenu/modmgr.lua]: Line too long

File: builtin/mainmenu/modmgr.lua [hunk 6]
```diff
@@ -248,35 +267,57 @@ function modmgr.render_modlist(render_list)
...
+			end
 		elseif v.enabled then
 			color = mt_color_green
 		end
 
 		retval[#retval + 1] = color
-		if v.modpack ~= nil or v.typ == "game_mod" then
-			retval[#retval + 1] = "1"
+		if v.typ == "game_mod" then
+			retval[#retval + 1] = tostring((v.mp_level or 0)+1)
```
Comment (SmallJoker) [inline builtin/mainmenu/modmgr.lua]: Spaces for readability: `or 0) + 1)`

File: mods/mods_here.txt (hunk 3 above)
Comment (orwell96) [inline mods/mods_here.txt]: That was accidental.

**Comment by orwell96** [comment]
Removing the mods_here.txt file was accidental, I symlinked the mods directory of the git repo to ~/.minetest/mods, where that file was missing and just accidentally added it to the commit.
Sorry for amending the commit so often, but you like tidy histories, so...

**Comment by orwell96** [comment]
Can someone please review this, I would like to see it in the release! Thanks!

**Comment by SmallJoker** [comment]
If there is already the possibility to have nested modpacks in the core, we also should allow them in the main menu. Sorry for the delay - I will review the code again once it's rebased.

**Comment by orwell96** [comment]
I can't rebase it right now, only next week.

**Comment by orwell96** [comment]
Theoretically, I have rebased this PR, but GitHub did not update the PR for some reason.
Look here:
https://github.com/orwell96/minetest/tree/nested_modpacks

**Comment by orwell96** [comment]
@SmallJoker?

**Comment by SmallJoker** [comment]
Nested modpacks work nicely but the subgame (+ its mods) are no longer shown. Please include them in this list again.

**Comment by orwell96** [comment]
This must have been caused by the conflicting change. I will look over it on monday.

**Comment by orwell96** [comment]
Finally got to fix the problem!
Now it works as expected. Also rebased again.

**Comment by orwell96** [comment]
Is someone still monitoring this PR?

**Comment by orwell96** [comment]
Thanks

**Comment by orwell96** [comment]
Ahem...

**Comment by orwell96** [comment]
Ahem...

**Comment by pgimeno** [comment]
@orwell96 Hi, sorry, I was AFK when you contacted me on IRC. Would you please contact me by email so we can discuss putting together the effort for this PR and #6898? Please refer to my homepage for an email address: http://www.formauri.es/personal/pgimeno/

**Comment by orwell96** [comment]
This PR is now too outdated to be applied, judging from the heavy changes to the mod management.
I will redo this and open a new PR.

**Pull request closed without merging**